File: 프로그래머스/lv3/42628. 이중우선순위큐/이중우선순위큐.java
import java.util.PriorityQueue;

class Solution {
    public int[] solution(String[] operations) {
        PriorityQueue<Integer> negPq = new PriorityQueue<>();
        PriorityQueue<Integer> pq = new PriorityQueue<>((o1, o2) -> o2 - o1);
        for (int i = 0; i < operations.length; i++) {
            String[] oper = operations[i].split(" ");
            if (oper[0].equals("I")) {
                pq.add(Integer.parseInt(oper[1]));
                negPq.add(Integer.parseInt(oper[1]));
            } else {
                if (oper[1].equals("-1")) {
                    if (!negPq.isEmpty()) {
                        negPq.poll();
                        if (pq.size() == 1) {
                            pq.poll();
                        }
                    }
                } else {
                    if (!pq.isEmpty()) {
                        pq.poll();
                        if (negPq.size() == 1) {
                            negPq.poll();
                        }
                    }
                }
            }
        }

        int max = 0, min = (int) 1e9;
        int num;

        if (pq.size() >= negPq.size()) {
            while (!negPq.isEmpty()) {
                num = negPq.poll();
                if (pq.contains(num)) {
                    max = Math.max(max, num);
                    min = Math.min(min, num);
                }
            }
        } else {
            while (!pq.isEmpty()) {
                num = pq.poll();
                if (negPq.contains(num)) {
                    max = Math.max(max, num);
                    min = Math.min(min, num); 
                }
            }
        }

        if (min == (int) 1e9) {
            min = 0;
        }
        int[] answer = {max, min};
        return answer;
    }
}

File: 프로그래머스/lv3/42628. 이중우선순위큐/이중우선순위큐.py
from heapq import heappop, heappush

def solution(operations):
    visited = [False] * 1000001
    max_heap = []
    min_heap = []
    for i in range(len(operations)):
        key, num = operations[i].split(" ")
        num = int(num)
        if key == "I":
            heappush(max_heap, (-num, i))
            heappush(min_heap, (num, i))
            visited[i] = True
        else:
            if num == -1:
                while min_heap and not visited[min_heap[0][1]]:
                    heappop(min_heap)
                if min_heap and visited[min_heap[0][1]]:
                    visited[min_heap[0][1]] = False
                    heappop(min_heap)
            else:
                while max_heap and not visited[max_heap[0][1]]:
                    heappop(max_heap)
                if max_heap and visited[max_heap[0][1]]:
                    visited[max_heap[0][1]] = False
                    heappop(max_heap)
                    
    while min_heap and not visited[min_heap[0][1]]:
        heappop(min_heap)
    while max_heap and not visited[max_heap[0][1]]:
        heappop(max_heap)
    if min_heap and max_heap:
        answer = [-max_heap[0][0], min_heap[0][0]]
    else:
        answer = [0, 0]
    return answer